### PR TITLE
support OIDC protocol

### LINKED ISSUE #3378: Feature: Support OIDC protocol
**What's it about?**

OpenID Connect 1.0 is a simple identity layer on top of the OAuth 2.0 protocol. It allows Clients to verify the identity of the End-User based on the authentication performed by an Authorization Server, as well as to obtain basic profile information about the End-User in an interoperable and REST-like manner.

OpenID Connect allows clients of all types, including Web-based, mobile, and JavaScript clients, to request and receive information about authenticated sessions and end-users. The specification suite is extensible, allowing participants to use optional features such as encryption of identity data, discovery of OpenID Providers, and session management, when it makes sense for them.

**What's the reason why we need it?**

The ks-core requires the ability to provide authentication to additional components, such as [Jenkins](https://plugins.jenkins.io/oic-auth/), [Gitlab](https://docs.gitlab.com/ee/administration/auth/oidc.html), [Harbor](https://goharbor.io/docs/1.10/administration/configure-authentication/oidc-auth/).

OAuth steps are illustrated in the following diagram:

```
+--------+                                   +--------+
|        |                                   |        |
|        |---------(1) AuthN Request-------->|        |
|        |                                   |        |
|        |  +--------+                       |        |
|        |  |        |                       |        |
|        |  |  End-  |<--(2) AuthN & AuthZ-->|        |
|        |  |  User  |                       |        |
|   RP   |  |        |                       |   KS   |
|        |  +--------+                       |        |
|        |                                   |        |
|        |<--------(3) AuthN Response--------|        |
|        |                                   |        |
|        |---------(4) UserInfo Request----->|        |
|        |                                   |        |
|        |<--------(5) UserInfo Response-----|        |
|        |                                   |        |
+--------+                                   +--------+

```

**TODO**

* [x] #4155
* [x] Implementing the [OIDC Core Protocol](https://openid.net/specs/openid-connect-core-1_0.html).
* [ ] https://github.com/kubesphere/console/issues/2208
* [ ] Add e2e test.

**Area Suggestion**

/area iam
/kind feature-request

### PR DESCRIPTION
**Pull request by wansir**
Signed-off-by: hongming <talonwan@yunify.com>

**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

KubeSphere will be used as an identity provider, the OIDC protocol can provide identity authentication capabilities for other components.

**Which issue(s) this PR fixes**:

Fixes #3378

**Special notes for reviewers**:

Depends on https://github.com/kubesphere/kubesphere/pull/4156

### Does this PR introduced a user-facing change?

```release-note
Built-in OAuth 2.0 authorization server with OpenID Connect protocol.
```

### Additional documentation, usage docs, etc.:

Additional documentation
1. [OIDC protocol](https://openid.net/specs/openid-connect-core-1_0.html)
2. [OAuth2](https://datatracker.ietf.org/doc/html/rfc6749)

Some request message:

Discovery

```
$ curl http://localhost:9090/oauth/.well-known/openid-configuration
{
 "issuer": "https://ks.wansir.com/oauth",
 "authorization_endpoint": "https://ks.wansir.com/oauth/authorize",
 "token_endpoint": "https://ks.wansir.com/oauth/token",
 "userinfo_endpoint": "https://ks.wansir.com/oauth/userinfo",
 "jwks_uri": "https://ks.wansir.com/oauth/keys",
 "grant_types_supported": [
  "authorization_code",
  "refresh_token"
 ],
 "response_types_supported": [
  "code",
  "token"
 ],
 "subject_types_supported": [
  "public"
 ],
 "id_token_signing_alg_values_supported": [
  "RS256"
 ],
 "code_challenge_methods_supported": [
  "S256",
  "plain"
 ],
 "scopes_supported": [
  "openid",
  "email",
  "profile",
  "offline_access"
 ],
 "token_endpoint_auth_methods_supported": [
  "client_secret_basic",
  "client_secret_post"
 ],
 "claims_supported": [
  "iss",
  "sub",
  "aud",
  "iat",
  "exp",
  "email",
  "locale",
  "preferred_username"
 ]
```

Authentication using the `Authorization Code Flow`: 
```
$ curl -u admin:P@88w0rd -X GET http://localhost:9090/oauth/authorize\?client_id\=default\&redirect_uri\=http://localhost:9091\&response_type\=code\&scope\=openid+profile+email -v
Note: Unnecessary use of -X or --request, GET is already inferred.
* Uses proxy env variable http_proxy == 'http://127.0.0.1:7890'
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 7890 (#0)
* Server auth using Basic with user 'admin'
> GET http://localhost:9090/oauth/authorize?client_id=default&redirect_uri=http://localhost:9091&response_type=code&scope=openid+profile+email HTTP/1.1
> Host: localhost:9090
> Authorization: Basic YWRtaW46UEA4OHcwcmQ=
> User-Agent: curl/7.64.1
> Accept: */*
> Proxy-Connection: Keep-Alive
>
< HTTP/1.1 302 Found
< Content-Length: 371
< Connection: keep-alive
< Content-Type: text/html; charset=utf-8
< Date: Tue, 14 Sep 2021 11:08:46 GMT
< Keep-Alive: timeout=4
< Location: http://localhost:9091?code=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOlsiZGVmYXVsdCJdLCJleHAiOjE2MzE2MTgzMjYsImlhdCI6MTYzMTYxNzcyNiwiaXNzIjoiaHR0cHM6Ly9rcy53YW5zaXIuY29tL29hdXRoIiwic3ViIjoiYWRtaW4iLCJ0b2tlbl90eXBlIjoiY29kZSIsInVzZXJuYW1lIjoiYWRtaW4iLCJzY29wZXMiOlsib3BlbmlkIiwicHJvZmlsZSIsImVtYWlsIl19.Doy7wRXcjNLAVw-E3cELW5oC6qydLaq7iemUHj4eK_Q
< Proxy-Connection: keep-alive
<
<a href="http://localhost:9091?code=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOlsiZGVmYXVsdCJdLCJleHAiOjE2MzE2MTgzMjYsImlhdCI6MTYzMTYxNzcyNiwiaXNzIjoiaHR0cHM6Ly9rcy53YW5zaXIuY29tL29hdXRoIiwic3ViIjoiYWRtaW4iLCJ0b2tlbl90eXBlIjoiY29kZSIsInVzZXJuYW1lIjoiYWRtaW4iLCJzY29wZXMiOlsib3BlbmlkIiwicHJvZmlsZSIsImVtYWlsIl19.Doy7wRXcjNLAVw-E3cELW5oC6qydLaq7iemUHj4eK_Q">Found</a>.
```

```
$ curl -X POST 'localhost:9090/oauth/token' -H 'Content-Type: application/x-www-form-urlencoded' --data-urlencode 'grant_type=code' --data-urlencode 'client_id=default' --data-urlencode 'client_secret=kubesphere' --data-urlencode 'code=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOlsiZGVmYXVsdCJdLCJleHAiOjE2MzE2MTgzMjYsImlhdCI6MTYzMTYxNzcyNiwiaXNzIjoiaHR0cHM6Ly9rcy53YW5zaXIuY29tL29hdXRoIiwic3ViIjoiYWRtaW4iLCJ0b2tlbl90eXBlIjoiY29kZSIsInVzZXJuYW1lIjoiYWRtaW4iLCJzY29wZXMiOlsib3BlbmlkIiwicHJvZmlsZSIsImVtYWlsIl19.Doy7wRXcjNLAVw-E3cELW5oC6qydLaq7iemUHj4eK_Q'
{
 "access_token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE2MzE2MjUxMDYsImlhdCI6MTYzMTYxNzkwNiwiaXNzIjoiaHR0cHM6Ly9rcy53YW5zaXIuY29tL29hdXRoIiwic3ViIjoiYWRtaW4iLCJ0b2tlbl90eXBlIjoiYWNjZXNzX3Rva2VuIiwidXNlcm5hbWUiOiJhZG1pbiJ9.aqwe9IM-PZoBNd34TcXztN34r-YkrjoiK61ZO6gX_A8",
 "token_type": "Bearer",
 "refresh_token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE2MzE2MzIzMDYsImlhdCI6MTYzMTYxNzkwNiwiaXNzIjoiaHR0cHM6Ly9rcy53YW5zaXIuY29tL29hdXRoIiwic3ViIjoiYWRtaW4iLCJ0b2tlbl90eXBlIjoicmVmcmVzaF90b2tlbiIsInVzZXJuYW1lIjoiYWRtaW4ifQ.EE-0DG013N8_5B93TWDy81YuOEx--EuG3EfDY962xzQ",
 "id_token": "eyJhbGciOiJSUzI1NiIsImtpZCI6IjU0OTgzNDQ2OSIsInR5cCI6IkpXVCJ9.eyJhdWQiOlsiZGVmYXVsdCJdLCJleHAiOjE2MzE2MzIzMDYsImlhdCI6MTYzMTYxNzkwNiwiaXNzIjoiaHR0cHM6Ly9rcy53YW5zaXIuY29tL29hdXRoIiwic3ViIjoiYWRtaW4iLCJ0b2tlbl90eXBlIjoiaWRfdG9rZW4iLCJ1c2VybmFtZSI6ImFkbWluIiwibmFtZSI6ImFkbWluIiwiZW1haWwiOiJhZG1pbkBrdWJlc3BoZXJlLmlvIiwibG9jYWxlIjoiemgiLCJwcmVmZXJyZWRfdXNlcm5hbWUiOiJhZG1pbiJ9.aRxai-q09Z2spUm6YfTklxmYZv5I8KVvdktajd9wFwLLGKb7EJ_yZkh1HK70IPHit5UiY0SXLkr3nUSHaqQnGylRjJL2yVWqQsaj8bBmBHfRPNfVI_I347TBGo9kEv0nd12GGufmSf5YvNyafyUAfWz1d_IVr0nFO6_hQ-Fi8REX18osXJRIb7-eTusvG6ZgcENcwbmDQAOXVeMKv69kuh4G0flYR9IY78kEtsfsBUyvflsnNfJriUetwqx9jkWm00vLbVHWjrkVMCCEhfI2qv1ukB3V2HwgJ0YlJZON9q9G5dsxQgnAZdO3hMpNFrOCk3rkF-2QoAbpY2JDQ7QHdw",
 "expires_in": 7200
}
```

### CODE REVIEW

File: pkg/apiserver/authentication/identityprovider/github/github_test.go [hunk 1]
```diff
@@ -102,7 +103,7 @@ scopes:
 					TokenURL:    tokenURL,
 					UserInfoURL: userInfoURL,
 				},
-				RedirectURL: "http://ks-console/oauth/redirect",
+				RedirectURL: "https://console.kubesphere.io/oauth/redirect/github",
```
Comment (zryfish) [inline pkg/apiserver/authentication/identityprovider/github/github_test.go:106]: Please don't use `https://console.kubesphere.io` in test code.

File: pkg/apiserver/authentication/oauth/error.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
+/*
+Copyright 2020 The KubeSphere Authors.
```
Comment (zryfish) [inline pkg/apiserver/authentication/oauth/error.go:2]: 2021

File: pkg/apiserver/authentication/oauth/error.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package oauth
+
+import "fmt"
+
+var (
+	// ErrorInvalidClient
```
Comment (zryfish) [inline pkg/apiserver/authentication/oauth/error.go:22]: Isn't this code from somewhere?

File: pkg/apiserver/authentication/identityprovider/github/github_test.go (hunk 1 above)
Comment (wansir) [reply pkg/apiserver/authentication/identityprovider/github/github_test.go]: fixed

File: pkg/apiserver/authentication/oauth/error.go (hunk 3 above)
Comment (wansir) [reply pkg/apiserver/authentication/oauth/error.go:23]: The following error type is defined in https://datatracker.ietf.org/doc/html/rfc6749#section-5.2

File: pkg/kapis/oauth/register.go [hunk 4]
```diff
@@ -50,42 +53,65 @@ func AddToContainer(c *restful.Container, im im.IdentityManagementInterface,
...
-		Consumes("application/x-www-form-urlencoded").
+		Consumes(contentTypeFormData).
 		Doc("The resource owner password credentials grant type is suitable in\n"+
 			"cases where the resource owner has a trust relationship with the\n"+
 			"client, such as the device operating system or a highly privileged application.").
 		Param(ws.FormParameter("grant_type", "Value MUST be set to \"password\".").Required(true)).
 		Param(ws.FormParameter("username", "The resource owner username.").Required(true)).
 		Param(ws.FormParameter("password", "The resource owner password.").Required(true)).
-		To(handler.Token).
+		To(handler.token).
```
Comment (RolandMa1986) [inline pkg/kapis/oauth/register.go:114]: Please add all the required `FormParameter`.

**Comment by zryfish** [comment]
Please fix conflicts, and please write a more detailed release-note

File: pkg/apiserver/authentication/oauth/error.go (hunk 3 above)
Comment (zryfish) [reply pkg/apiserver/authentication/oauth/error.go:23]: Best to point out, include their license would be great

Comment (wansir) [reply pkg/apiserver/authentication/oauth/error.go:23]: Already added the external doc link.

File: pkg/kapis/oauth/register.go (hunk 4 above)
Comment (wansir) [reply pkg/kapis/oauth/register.go:119]: done

**Comment by wansir** [comment]
> Please fix conflicts, and please write a more detailed release-note

done

**Pull request merged**